bin/propolis-server/src/lib/vm/ensure.rs

```diff
@@ -0,0 +1,359 @@
+                    log: self.log,
+                    vm: self.vm,
+                    ensure_request: self.ensure_request,
+                    ensure_options: self.ensure_options,
+                    ensure_response_tx: self.ensure_response_tx,
+                    state_publisher: self.state_publisher,
+                    vm_objects: objects,
+                    input_queue,
+                    kernel_vm_paused: false,
+                })
+            }
+            Err(e) => Err(self.fail(e).await),
+        }
+    }
+
+    pub(crate) async fn fail(self, reason: anyhow::Error) -> anyhow::Error {
+        self.state_publisher
+            .update(ExternalStateUpdate::Instance(InstanceState::Failed));
+
+        self.vm.vm_init_failed().await;
+        let _ = self
+            .ensure_response_tx
+            .send(Err(VmError::InitializationFailed(reason.to_string())));
+
+        reason
+    }
+}
+
+/// Represents an instance ensure request that has proceeded far enough to
+/// create a set of VM objects, but that has not yet installed those objects as
+/// an `ActiveVm` or notified the requestor that its request is complete.
+pub(crate) struct VmEnsureObjectsCreated<'a> {
+    log: &'a slog::Logger,
+    vm: &'a Arc<super::Vm>,
+    ensure_request: &'a InstanceSpecEnsureRequest,
+    ensure_options: &'a EnsureOptions,
+    ensure_response_tx: InstanceEnsureResponseTx,
+    state_publisher: &'a mut StatePublisher,
+    vm_objects: Arc<VmObjects>,
+    input_queue: Arc<InputQueue>,
+    kernel_vm_paused: bool,
+}
+
+impl<'a> VmEnsureObjectsCreated<'a> {
+    /// Prepares the VM's CPUs for an incoming live migration by activating them
+    /// (at the kernel VM level) and then pausing the kernel VM. This must be
+    /// done before importing any state into these objects.
+    ///
+    /// # Panics
+    ///
+    /// Panics if called more than once on the same set of objects.
+    pub(crate) async fn prepare_for_migration(&mut self) {
+        assert!(!self.kernel_vm_paused);
+        let guard = self.vm_objects.lock_exclusive().await;
+        guard.reset_vcpus();
+        guard.pause_kernel_vm();
+        self.kernel_vm_paused = true;
+    }
+
+    /// Uses this struct's VM objects to create a set of VM services, then
+    /// installs an active VM into the parent VM state machine and notifies the
+    /// ensure requester that its request is complete.
+    pub(crate) async fn ensure_active(self) -> VmEnsureActive<'a> {
+        let vm_services = VmServices::new(
+            self.log,
+            self.vm,
+            &self.vm_objects,
+            &self.ensure_request.properties,
+            self.ensure_options,
+        )
+        .await;
+
+        self.vm
+            .make_active(
+                self.log,
+                self.input_queue.clone(),
+                &self.vm_objects,
+                vm_services,
+            )
+            .await;
+
+        let _ = self.ensure_response_tx.send(Ok(InstanceEnsureResponse {
```

Comment:
what does it actually mean if the receiver has gone away, here? should we give up on the ensure?